File: 373.find-k-pairs-with-smallest-sums.py
#
# @lc app=leetcode id=373 lang=python3
#
# [373] Find K Pairs with Smallest Sums
#

# Time:  O(k * log(m, n, k))
# Space: O(min(m, n, k))

import heapq
from typing import List


# @lc code=start
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        if len(nums1) > len(nums2):
            self.kSmallestPairs(nums2, nums1, k)

        pairs, min_heap = [], []

        def push(i, j):
            if i < len(nums1) and j < len(nums2):
                heapq.heappush(min_heap, [nums1[i] + nums2[j], i, j])

        push(0, 0)
        while min_heap and len(pairs) < k:
            _, i, j = heapq.heappop(min_heap)
            pairs.append([nums1[i], nums2[j]])
            push(i, j + 1)
            if j == 0:
                push(i + 1, 0)  # at most queue min(n, m) space
        return pairs


# @lc code=end

File: 373.find-k-pairs-with-smallest-sums.js
/*
 * @lc app=leetcode id=373 lang=javascript
 *
 * [373] Find K Pairs with Smallest Sums
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number[][]}
 */
var kSmallestPairs = function (nums1, nums2, k) {
  const res = [];
  for (let i = 0; i < Math.min(nums1.length, k); ++i) {
    for (let j = 0; j < Math.min(nums2.length, k); ++j) {
      res.push([nums1[i], nums2[j]]);
    }
  }
  res.sort((a, b) => a[0] + a[1] - b[0] - b[1]);
  return res.length > k ? res.slice(0, k) : res;
};
// @lc code=end
